### PR TITLE
Vg pdf jung kurt

### PR DESCRIPTION
**Pull request by sbinet**

### CODE REVIEW

**Comment by sbinet** [comment]
the introduction of `makeFont` (`vg/vgpdf/vgpdf.go`) is somewhat unfortunate but necessary.

I'll create an issue against the gofpdf repo to ask for an API that isn't file-oriented (but also has the ability to take an `io.Reader`)

**Comment by sbinet** [comment]
ping?

File: vg/vgpdf/vgpdf.go [hunk 1]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
 // Canvas implements the vg.Canvas interface,
 // drawing to a PDF.
 type Canvas struct {
-	doc         *pdf.Document
+	pdf         *pdf.Fpdf
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: IIRC this was mainly to detect all places that needed to be changed (and not happen to have the same signature)
will change back to `doc` as a field name.

File: vg/vgpdf/vgpdf.go [hunk 2]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
 	w, h        vg.Length
-	page        *pdf.Canvas
 	lineVisible bool
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: ok.

File: vg/vgpdf/vgpdf.go [hunk 3]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
+
+	dpi   int
+	nimgs int // images counter
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: ok

File: vg/vgpdf/vgpdf.go [hunk 4]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
+	ctx   []context
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: ok

File: vg/vgpdf/vgpdf.go [hunk 5]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
+	fonts map[vg.Font]struct{}
+}
+
+type context struct {
+	fillc color.Color // fill color
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: ok

File: vg/vgpdf/vgpdf.go [hunk 6]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
+	c.Scale(1, -1)
+
 	return c
 }
 
+func (c *Canvas) DPI() float64 {
+	return float64(c.dpi)
+}
+
+func (c *Canvas) cur() *context {
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: ok

File: vg/vgpdf/vgpdf.go [hunk 7]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
+		// not defined (standard font?), use average of 81%
+		top = 0.81 * h
+	} else {
+		top = -float64(d.Ascent) * h / float64(d.Ascent-d.Descent)
+	}
+	return 0, top, c.pdf.GetStringWidth(txt), top + h
+}
+
+// whitespace determines if a rune is whitespace
+func whitespace(r rune) bool {
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: I was a tad reluctant to import the rather large `unicode` package just for that one function.
but ok.

File: vg/vgpdf/vgpdf.go [hunk 8]
```diff
@@ -25,123 +32,238 @@ const DPI = 72
...
+		if err != nil {
+			log.Panicf("vgpdf: could not generate font data for PDF: %v", err)
+		}
+
+		c.fonts[fnt] = struct{}{}
+		c.pdf.AddFontFromBytes(fnt.Name(), "", jdata, zdata)
+		return
 	}
-	c.page.DrawImage(img, r)
+	log.Panicf("vgpdf: not implemented")
```
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: sure :)

File: vg/vgpdf/vgpdf.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: (`vgsvg.Canvas` and `vgeps.Canvas` call it `stk`... `vgtex.Canvas` does call it `stack`)

**Comment by sbinet** [comment]
Done.

File: vg/vgpdf/vgpdf.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply vg/vgpdf/vgpdf.go]: If you want you can copy-paste the function from unicode, but the current code does not detect all whitespace runes.

File: vg/vgpdf/vgpdf.go [hunk 9]
```diff
@@ -25,123 +32,230 @@ const DPI = 72
...
+	w, h vg.Length
+
+	dpi       int
+	numImages int
+	stack     []context
+	fonts     map[vg.Font]struct{}
+}
+
+type context struct {
+	fill  color.Color // fill color
```
Comment (kortschak) [inline vg/vgpdf/vgpdf.go]: The comments don't need to exist anymore.

File: vg/vgpdf/vgpdf.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: tunrs out it was a leftover from some brownian-walk code production...
case closed.

File: vg/vgpdf/vgpdf.go (hunk 9 above)
Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: Done.

**Comment by sbinet** [comment]
PTAL

File: vg/vgpdf/vgpdf.go [hunk 10]
```diff
@@ -229,12 +343,74 @@ func (w *writerCounter) Write(p []byte) (int, error) {
...
+	c255 = 255.0 / 65535.0
+)
+
+// rgba converts a Go color into a gofpdf 3-tuple int + 1 float64
+func rgba(c color.Color) (int, int, int, float64) {
+	if c == nil {
+		c = color.Black
+	}
+	r, g, b, a := c.RGBA()
+	return int(float64(r) * c255), int(float64(g) * c255), int(float64(b) * c255), float64(a) / math.MaxUint16
```
Comment (kortschak) [inline vg/vgpdf/vgpdf.go]: Shouldn't these just be `x>>8`?

**Comment by sbinet** [comment]
PTAL.

**Comment by sbinet** [comment]
(this is just a rebase + nice-ification of temporary CLs during the review)

should I wait a bit for @eaburns or do I go ahead (and get rid of our last `hg` build dependency) ?

**Comment by sbinet** [comment]
ok. I've waited long enough :)

**Pull request merged**